Implement initial support for VP9 decoding in LibVideo

Zaggy1024 opened the pull request:
This completes the initial implementation of the LibVideo VP9 decoder, building off the work of FalseHonesty. I'm glad most of the work on parsing was already done. :^)

There's a lot more work to do on this to make it user-friendly, but VideoPlayer can open WebM video files and display them, and with a left click it can advance one frame. It's only for testing at the moment, and just reads the internal Vector<u16> output buffer and then converts that from YCbCr to RGB for an image widget.

The decoder still needs to implement loop filtering and a proper output function that bundles up a whole frame for the consumer in a helpful manner.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/BitStream.cpp:
```diff
@@ -82,61 +108,85 @@ bool BitStream::read_bool(u8 probability)
-        }
-        m_bool_range *= 2;
-        m_bool_value = (m_bool_value << 1u) + new_bit;
+    // Number of bits to read to shift the range above 128. Calculated by 7 - log2(n).
+    constexpr u8 range_shifts[127] = {
+        // 1..32
+        7, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2,
+        // 33..64
+        2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1,
+        // 65..96
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        // 97..127
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
+    };
```
Is this some kind of optimization? If so, it could also be split into a seperate commit?

MacDue replied:
Optional: A kinda neat trick for things like this to compute the values at compile time rather than hardcode the table. 
I believe something like this would work here (you may need to tweak it a little).
```c++
constexpr auto range_shifts = [] {
    Array<u8, 127> lut {};
    for (size_t n = 1; n <= lut.size(); n++)
        lut[n-1] = static_cast<u8>(7 - AK::log2(n));
    return lut;
}();
```

Zaggy1024 replied:
I've cleaned up the constexpr initialization as you suggested, thanks!

FireFox317 commented on Userland/Libraries/LibVideo/VP9/Parser.cpp:
```diff
@@ -1399,4 +1410,8 @@ void Parser::dump_info()
 
+#undef RESERVED_ZERO
+#undef FUNC_ERROR
+#undef WRAP_ERROR
```
What's up with these undefs?

Zaggy1024 replied:
My bad, those are leftover from before. (Also, I was apparently writing the file thinking of it as a header, so that was pointless anyway.)

FireFox317 commented on Userland/Libraries/LibVideo/VP9/Utilities.h:
```diff
@@ -1,22 +1,24 @@
+    auto value_cast = static_cast<OutputType>(value);
+    VERIFY(value == static_cast<InputType>(value_cast));
+    return value_cast;
+}
```
Can't we use something in AK/TypeCasts.h for this?

Zaggy1024 replied:
Now that I think about it, this function was unnecessary in light of my later discovery of the Clang integer sanitizer, so I've removed this.

FireFox317 commented on Userland/Libraries/LibVideo/CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
 set(SOURCES
+    DecoderError.h
```
Header files should not be added to the SOURCES set. This basically instructs CMake to build these header files, which is not something we want.

Might be a good idea to also remove all the other headers in this file.

ADKaster replied:
Actually it doesn't. IDEs like QtCreator use headers in the sources list (or target_headers or w/e) to populate the project.

but correct, we don't normally put headers in SOURCES.

Zaggy1024 replied:
Yeah, here I made the mistake of following what was already in the file.. I've added a first commit to remove the existing ones and fixed the added one.

FireFox317 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,122 @@
+using DecoderErrorOr = ErrorOr<T, DecoderError>;
+
+enum class DecoderErrorCategory : u32 {
+    Unknown = 0,
```
Enums start with 0 by default, so there is no need for the `= 0`.

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,122 @@
+        return DecoderError::with_description(category, String::vformatted(format_string.view(), variadic_format_params));
+    }
+
+    DecoderErrorCategory category()
+    {
+        return m_category;
+    }
+
+    StringView description()
+    {
+        return m_description;
+    }
+
+    StringView string_literal()
+    {
+        return m_description;
+    }
```
Simple functions like these can be put on a single line.

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    DecoderErrorCategory m_category { DecoderErrorCategory::Unknown };
+    String m_description { String::empty() };
```
Isn't using the default constructor here better?
```suggestion
    String m_description;
```

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,122 @@
+    String m_description { String::empty() };
+};
+
+// VSCode doesn't know __FILE_NAME__ exists
+#ifndef __FILE_NAME__
+#    define __FILE_NAME__ ""
+#endif
```
Is this code here just to make your IDE happy? If it builds without this, please remove it.

Zaggy1024 replied:
Yeah, without it I had errors in every TRY statement in Parser.cpp. I've changed __FILE_NAME__ usage to __FILE__ instead, though that will be a lot more verbose.

ADKaster replied:
Sorry to comment on a closed item, but looking at the code here, I think you should embrace our lord and savior AK::SourceLocation and its magic SourceLocation::current() value instead of relying on these macros to create descriptive errors.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/Decoder.h:
```diff
@@ -10,6 +10,9 @@
 #include <AK/Error.h>
+#include <LibVideo/DecoderError.h>
+
+#include "Parser.h"
```
Uhm, Parser.h is already include a few lines above?

Zaggy1024 replied:
Ah, probably goofed up in a rebase. Fixed.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/Parser.cpp:
```diff
@@ -12,6 +12,8 @@
 
 namespace Video::VP9 {
 
+#define TRY_READ(expression) DECODER_TRY(DecoderErrorCategory::Corrupted, expression)
```
Why is this not defined in the `DecoderError.h` file?

Zaggy1024 replied:
Because this is only relevant during the parsing process, so it seemed pointless to make it accessible to other files. I have a few Category::Corrupted errors in Decoder.cpp, but not ones that would use the TRY syntax, so it wasn't applicable there. This macro is just for convenience to make the hundreds of try calls less verbose in Parser.

Should I move it anyway?

Zaggy1024 replied:
I'll leave this as is and resolve this conversation.

FireFox317 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -21,6 +21,7 @@ using DecoderErrorOr = ErrorOr<T, DecoderError>;
 enum class DecoderErrorCategory : u32 {
     Unknown = 0,
     IO,
+    Memory,
```
This could go in the previous commit? Or it might even be nicer to have a complete separate commit that adds the `DecoderError.h` file, and then in a subsequent commit actually use the new functionality.

Zaggy1024 replied:
Hmm, I was thinking it made sense to add the classes of errors as they're used so that you could see their usage in the same commit.

I've added a few more "error" types (some are not exactly errors for decoders) since this PR, and I'm not sure if it may be necessary to add some more on top.

I could definitely split this up into multiple commits if necessary, though.

Zaggy1024 replied:
Leaving this as-is if that's all right.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/BitStream.h:
```diff
@@ -38,11 +38,11 @@ class BitStream {
     u64 get_position();
-    size_t bytes_remaining();
     size_t bits_remaining();
+    size_t bytes_remaining();
```
This change seems unnecessary? It just moves the definition 2 lines downward.

Zaggy1024 replied:
Ah, this was just to match the order in the cpp file so that it makes sense since one references the other. I'll revert this.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/BitStream.cpp:
```diff
@@ -167,9 +167,9 @@ ErrorOr<u8> BitStream::read_literal(u8 bit_count)
+ErrorOr<i16> BitStream::read_s(size_t bits)
 {
-    auto value = TRY(read_bits(bits));
+    i16 value = TRY(read_bits(bits));
```
This is change is not explained in the commit message, and could belong in a separate commit.

Zaggy1024 replied:
It looks like this change may not have been necessary after all, so I'll revert it for now.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/Parser.h:
```diff
@@ -228,19 +242,18 @@ class Parser {
+    Vector<InterpolationFilter> m_interp_filters {};
+    Vector<Array<MV, 2>> m_mvs {};
+    Vector<Array<Array<MV, 4>, 2>> m_sub_mvs {};
+    Vector<Array<IntraMode, 4>> m_sub_modes {};
```
Default constructors don't need the `{}`.
```suggestion
    Vector<Array<IntraMode, 4>> m_sub_modes;
```

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -26,32 +26,75 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+                //        Break out of all loops so that we can display the first frame.
+                //        This loop should also be moved to something that stores demuxer state
+                //        and gets called from the GUI event loop.
+                goto display;
```
Please don't use `goto` in Serenity.

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Libraries/LibVideo/VP9/MotionVector.h:
```diff
@@ -0,0 +1,69 @@
+/*
```
The MotionVector change could be split into a separate commit.

Zaggy1024 replied:
I've split this commit up into:

1. Rename MV->MotionVector
2. Make MotionVector.h allow constexpr and create MotionVector tables
3. The rest of the inter frame parsing commit
4. Some bug fixes on old code that were included in the big inter frame commit previously

FireFox317 commented on Userland/Libraries/LibVideo/VP9/TreeParser.cpp:
```diff
@@ -665,80 +665,86 @@ u8 TreeParser::calculate_token_probability(u8 node)
 
 void TreeParser::count_syntax_element(SyntaxElementType type, int value)
 {
+#define INCREMENT_ELEMENT(element)                                    \
```
Instead of adding a macro for this can't we just use `AK::clamp`?

Zaggy1024 replied:
I changed it to use a local lambda to set `x = min(x + 1, 255)` for each element instead, and reworded the commit message since clamp is invalid for unsigned integers.

FireFox317 commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -28,72 +45,98 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+        auto total_milliseconds = (seconds * 1000) + (useconds / 1000);
+        outln("Decoding 100 frames took {} ms", total_milliseconds);
+        return 0;
+    }
```
You could use `ElapsedTimer` from `LibCore` instead of doing this the manual way with libc.

Zaggy1024 replied:
Fixed.

FireFox317 commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -13,13 +13,30 @@
+    bool benchmark = false;
+    StringView filename = "/home/anon/Videos/test-webm.webm"sv;
+
+    // FIXME: Very lazy argument parsing for testing purposes
```
Maybe use `ArgsParser`?

Zaggy1024 replied:
Fixed.

FireFox317 reviewed (commented):
In general I think commits can be split into smaller atomic commits, and make it easier to review that way. It's not like you are not allowed to have a lot of commits, this is a significant change anyway. I tried to give you some useful feedback nevertheless :^)

Zaggy1024 commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -43,10 +43,12 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
```
I've split up that commit as suggested.

Zaggy1024 commented:
I've finished updating this based on your comments, and split up several of the commits into more pieces. All of them should be self-contained changes that still compile, so hopefully it all helps make this more digestible.

Let me know if there are any other commits I should split up.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Decoder.h:
```diff
@@ -33,16 +46,101 @@ class Decoder {
+    inline i32 sin64(u8 angle);
+    /* The function B( a, b, angle, 0 ) performs a butterfly rotation. */
+    inline void butterfly_rotation_in_place(Vector<Intermediate>& data, size_t index_a, size_t index_b, u8 angle, bool flip);
+    /* The function H( a, b, 0 ) performs a Hadamard rotation. */
```
This file has a mix of `//` and `/* */` style comments, which seems to be somewhat pre-existing? 

However in commit `LibVideo: Implement VP9 intra-predicted frame decoding` you converted some to `//` and added a ton of `/* */` comments. Could we stick to new comments (outside of the file copyright header, for uh... reasons) being `//` and come back later (like a later PR) to fully clean up the c style comments that were pre-existing (when you don't touch that function anyway)?

Zaggy1024 replied:
Fixed those, turns out I had a few more /* laying around as well that I've fixed.

ADKaster commented on Userland/Libraries/LibVideo/VP9/MV.h:
```diff
@@ -13,6 +13,7 @@ namespace Video::VP9 {
 class MV {
 public:
     MV() = default;
+    MV(MV const& other);
```
this and operator=(MV const&) should be `= default;` in the header here. As well as defaulting a move constructor and move assign. Rule of 5 and all that.

Zaggy1024 replied:
I'll just remove that line, since I'm not sure how that snuck in there considering it's not necessary.

Zaggy1024 replied:
With regard to the rule of 5, I'll add the other functions if needed in my constexpr refactor of MotionVector.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Parser.h:
```diff
@@ -184,17 +192,23 @@ class Parser {
+    /// indicates that there are no transform coefficients.
+    ///
+    /// skip may be set to 0 even if transform blocks contain immediate end of block markers.
     bool m_skip { false };
```
m_skip seems a candidate for conversion to a verbose enum

```c++
enum TransformCoefficientsPossible
{
     Yes = 0,
     No = 1,
};
```

or something? probably a follow-up, but the lengthy comment here suggests that `bool` is not the right data type for this.

Zaggy1024 replied:
I've added a FIXME for this. Definitely another candidate for later readability refactoring.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Utilities.cpp:
```diff
@@ -8,14 +8,4 @@
 
 namespace Video::VP9 {
 
-u8 clip_3(u8 x, u8 y, u8 z)
```
In commit `LibVideo: Implement VP9 intra-predicted frame decoding` Utilities.cpp becomes completely empty. Do more utilities find their way here later? if not, the file should be deleted in that commit.

Zaggy1024 replied:
My fault, I had meant to delete the file when I committed my changes. Fixed.

ADKaster commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -25,32 +25,84 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+                break;
         }
+
+        if (decoded_frame)
```
this heavily nested loop construct with the `if (decoded_frame) break;;` chain at the end feels like it would be much nicer to read if it was moved into a helper function where you could early instead of cascading the success result out of each loop.

Zaggy1024 replied:
I'll just copy over my later implementation of this since it's cleaner, no point in refactoring this since it was only hacked together to get a frame to display.

ADKaster commented on Userland/Libraries/LibVideo/VP9/MotionVector.h:
```diff
@@ -10,24 +11,60 @@
+        , m_column(col)
+    {
+    }
+    constexpr MotionVector(MotionVector const& other)
```
in case the commits get lost/shuffled, MotionVector (old MV) should have = default'd special member functions.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Parser.h:
```diff
@@ -119,16 +121,25 @@ class Parser {
+    MotionVector clamp_mv(MotionVector mvec, i32 border);
     size_t get_image_index(u32 row, u32 column);
+    void get_block_mv(u32 candidate_row, u32 candidate_column, u8 ref_list, bool use_prev);
+    void if_same_ref_frame_add_mv(u32 candidate_row, u32 candidate_column, ReferenceFrame ref_frame, bool use_prev);
+    void if_diff_ref_frame_add_mv(u32 candidate_row, u32 candidate_column, ReferenceFrame ref_frame, bool use_prev);
```
these two function names feel out of style

It's a bit sad that the old function names were using so many abbreviations. would be nice to replace all the `mv` with `motion_vector` at some point (after this PR)

But for these two in particular, they should be something more like
`add_mv_if_ref_frame_same`

to keep the verb towards the front of the method name.

Zaggy1024 replied:
For all the initial implementations, I was using the exact naming from the spec (converting to snake case or camel case where needed), what are your thoughts on leaving this for a later refactor?

I had planned on going through and cleaning up the decoder to make it more readable and fix up unnecessarily stored state once it was fully functional, to avoid uprooting the existing implementation and breaking things.

ADKaster replied:
yah this is fine for a later refactor.

ADKaster commented on Userland/Libraries/LibVideo/VP9/TreeParser.h:
```diff
@@ -29,8 +30,17 @@ class TreeParser {
 
-        TreeSelection(int const* values);
-        TreeSelection(int value);
+        constexpr TreeSelection(int const* values)
```
I might be missing something but the fact that this class doesn't know how many elements are in this array weirds me out.

yakplus on this one I guess.

Zaggy1024 replied:
Yeah, might be worth coming back to this, that's how it was before. Not sure if the class was even necessary here since trees with single values can just be decoded with a different function call, that's what other decoders do. Another thing to refactor, I suppose.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Decoder.cpp:
```diff
@@ -560,10 +560,328 @@ DecoderErrorOr<void> Decoder::predict_intra(u8 plane, u32 x, u32 y, bool have_le
 }
 
-DecoderErrorOr<void> Decoder::predict_inter(u8, u32, u32, u32, u32, u32)
+MotionVector Decoder::motion_vector_selection(u8 plane, u8 ref_list, u32 block_index)
```
`select_motion_vector` ? you converted "the motion vector clamping process" to `clamp_motion_vector` below. either way they should be consistent.

Zaggy1024 replied:
Ah, good point, I didn't notice the inconsistency when I was writing this. It helps that the section name is commented so it can still be found in spec.

That also reminds me, clamp_motion_vector is named very similarly to clamp_mv (spec name), I'll add a FIXME to rename later if it makes sense.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Decoder.cpp:
```diff
@@ -560,10 +560,328 @@ DecoderErrorOr<void> Decoder::predict_intra(u8 plane, u32 x, u32 y, bool have_le
+            block_buffer_at(row, column) = reference_frame_buffer_at(clamp(y + row, 0, reference_frame_buffer.size() / reference_frame_width - 1), clamp(x + column, 0, reference_frame_width - 1));
+        }
+    }
+    return {};*/
```
commented out code in commit `LibVideo: Implement inter prediction`, pls remove

ADKaster commented on Userland/Libraries/LibVideo/VP9/Parser.h:
```diff
@@ -7,8 +7,10 @@
+#include <AK/Array.h>
 #include <AK/OwnPtr.h>
 #include <AK/Span.h>
+#include <AK/Vector.h>
```
Adding Vector and Array to Parser.h in `LibVideo: Make probability tables save to the specified index` feels out of place

Zaggy1024 replied:
Looks like this was because ProbabilitiesTable.h was already including Vector.h, I'll move the inclusion of those to the inter prediction commit where more Vectors and Arrays were added to Parser.h.

ADKaster commented on Userland/Libraries/LibVideo/VP9/Decoder.cpp:
```diff
@@ -807,14 +806,6 @@ DecoderErrorOr<void> Decoder::predict_inter_block(u8 plane, u8 ref_list, u32 x,
         return block_buffer[row * width + column];
     };
 
-    /*for (auto row = 0u; row < height; row++) {
```
see other comment for ref but this should be deleted ~5 commits earlier so that it's never even added :) 

(commenting on `LibVideo: Remove unnecessary dbgln calls`).

ADKaster commented on Userland/Libraries/LibVideo/VP9/Parser.cpp:
```diff
@@ -1736,6 +1729,7 @@ void Parser::dump_info()
     outln("Frame dimensions: {}x{}", m_frame_width, m_frame_height);
     outln("Render dimensions: {}x{}", m_render_width, m_render_height);
     outln("Bit depth: {}", m_bit_depth);
+    outln("Show frame: {}", m_show_frame);
```
I feel like this was removed super early in the commit stack and just added back now in `LibVideo: Remove unnecessary dbgln calls` ? if so, the addition should time travel back to the commit that removed it.

Zaggy1024 replied:
Actually, the removed line previously was printing the interpolation filter, which was uninitialized memory for the keyframe that VideoPlayer would try to decode before my changes. I previously merged this change into my intra prediction commit. I'll split this off so it's clearer why I removed it:

https://github.com/SerenityOS/serenity/commit/4173f3427223aadb9d376e5427bd07e6dc53d962#diff-802bcbd4f9ce71bcc5674a95fea292fc8c4b2a98a8476dc1cfad4eae09073ad7L1424

ADKaster commented on Userland/Applications/VideoPlayer/main.cpp:
```diff
@@ -27,81 +42,101 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-    main_widget->set_layout<GUI::VerticalBoxLayout>();
-    auto image_widget = TRY(main_widget->try_add<GUI::ImageWidget>());
-    image_widget->set_bitmap(image);
     image_widget->set_fixed_size(video_track.pixel_width, video_track.pixel_height);
```
The widget code here is moved up, and down, and back up throughout the commit stack. can we only move them once? it makes it hard to tell what has changed in `VideoPlayer: Allow display of multiple frames by clicking the image`

Zaggy1024 replied:
This should be fixed now after updating the image display code.

ADKaster commented on Tests/LibVideo/TestVP9Decode.cpp:
```diff
@@ -0,0 +1,37 @@
+
+TEST_CASE(webm_in_vp9)
+{
+    auto matroska_document = Video::MatroskaReader::MatroskaReader::parse_matroska_from_file("/usr/Tests/LibVideo/vp9_in_webm.webm"sv);
```
This should be able to run in Lagom as well as on serenity, so in the future we should add the "relative file path" hax that other tests that read test inputs do.

Zaggy1024 replied:
Is this a relatively simple fix? I could modify this test case if there's a good example to reference. I didn't see anything of the sort when I was looking through other test files. :(

ADKaster replied:
Since you've installed the files into `/usr/Tests/LibVideo` (`install(FILES vp9_in_webm.webm DESTINATION usr/Tests/LibVideo)`) they will be available at "./vp9_in_webm.webm". Userland/Utilities/run-tests.cpp will chdir to the parent directory of your test before running it. We can use the same strategy in a future addition to Lagom/CMakeLists.txt to use a relative path there. Some other test files use a "try the abspath then try ./" approach, which could probably use some cleaning up.

So the action here is either to make it "./vp9_in_webm.webm",  or keep that in mind for future on-linux testing approaches.

Zaggy1024 replied:
I've made the change to the relative path. Unfortunately I wasn't able to confirm on my machine (using WSL2) that the Lagom tests run since it was encountering a compile error in Zlib.cpp when I run `./Meta/serenity.sh test lagom`:

```/home/zaggy/serenity/Userland/Libraries/LibCompress/Zlib.cpp: In member function ‘void Compress::ZlibCompressor::write_header(Compress::ZlibCompressionMethod, Compress::ZlibCompressionLevel)’:
/home/zaggy/serenity/Userland/Libraries/LibCompress/Zlib.cpp:94:31: error: cannot convert ‘Compress::ZlibCompressionMethod’ to ‘unsigned char:4’ in initialization
   94 |         .compression_method = compression_method,
      |                               ^~~~~~~~~~~~~~~~~~
      |                               |
      |                               Compress::ZlibCompressionMethod
```

I'm unsure yet whether this issue is still present in the latest revision, since this branch is quite outdated now.

ADKaster reviewed (changes requested):
This was quite the refactor you did! A lot of the old code was very C, so this is really nice to see. 

I've got some comments on style and patterns, and a bunch on commit hygiene. I suggest going through the commits one by one to see what I mean with some of the comments. github lets you do this pretty easily using the Commits tab with the arrow buttons in the top right of that view.

I also left some future-looking comments, feel free to defer them to an issue if you like, or just keep them on your personal yak-stack.

davidot commented on Userland/Libraries/LibVideo/VP9/Decoder.h:
```diff
@@ -138,9 +153,16 @@ class Decoder {
         Vector<u16> output[3];
     } m_buffers;
+
+    bool m_enable_debug { false };
```
Is this (still) used?

Zaggy1024 replied:
Ah, I guess I didn't manage to avoid staging this field one of the many times I had to add debug printing. Good catch, I've removed that.

davidot commented on Userland/Libraries/LibVideo/VP9/Parser.cpp:
```diff
@@ -28,8 +28,53 @@ Parser::~Parser()
+    // Reading from span data will be quicker than spinning up a BitStream.
+    u8 superframe_byte = frame_data[frame_data.size() - 1];
+    u8 superframe_marker = superframe_byte & 0b1110'0000;
+    if (superframe_marker == 0b1100'0000) {
```
You could move the comment `// 1. parsing the final byte of the chunk and checking that the superframe_marker equals 0b110,` to just above this if to make it clearer

Also the comment says you check for `0b110` but you check for `0b1100'0000` so maybe add a NOTE why this is different from the comment?

Zaggy1024 replied:
Good points, I'll make those changes.

davidot reviewed (commented):
First off, great work!
It's clear that this is quite a lot of code but that is just required for things like this. And the code looks quite clean for a first PR.

Just quickly looked through for anything that that stuck out to me.
In general this is a lot of code to add at once so a proper review (actually looking if all the code does what the spec/comments say) is much harder at this scale, but I assume we'll find anything during actual usage!

(I just saw Andrew also reviewed this so sorry for any double comments)

Zaggy1024 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,106 @@
+    String m_description;
+};
+
+#define DECODER_ERROR_FORMAT "{}() (" __FILE__ ":{}): "
+#define DECODER_ERROR_PARAMS __func__, __LINE__
```
Ah, here I was assuming that macros were the only way to do this. That's a pretty cool hack, thanks for pointing me to that!

Zaggy1024 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,106 @@
+            "{} not implemented", __func__);      \
+    })
+
+#define DECODER_ERROR_CORRUPTED(description)        \
```
I believe this is resolved by changing these into static functions using SourceLocation, correct?

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/BitStream.h:
```diff
@@ -16,15 +17,16 @@ class BitStream {
     ErrorOr<bool> read_bit();
 
+    ErrorOr<u64> read_bits(u8 bit_count);
+    template<typename T>
+    ErrorOr<T> read_integer();
```
Now that I think about it, I should have removed this function when I added read_bits. I'll do that now.

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/BitStream.cpp:
```diff
@@ -142,13 +142,15 @@ size_t BitStream::range_coding_bits_remaining()
-    u8 return_value = 0;
-    for (size_t i = 0; i < n; i++) {
-        return_value = (2 * return_value) + TRY(read_bool(128));
+    if (bits > sizeof(u8) * 8)
+        return Error::from_string_literal("Range decoder read is too large");
+    u64 result = 0;
```
Hmm, you're right, not sure why I changed the return type initially. I would guess I was trying to fix some ubsan issue, but judging by what this is doing, it wasn't fixing anything. I'll drop this change as the bit shifting also probably doesn't result in any tangible benefit.

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/BitStream.cpp:
```diff
@@ -105,16 +107,24 @@ ErrorOr<bool> BitStream::read_bool(u8 probability)
+    }();
+
+    if (m_bool_range < 128) {
+        auto shift = range_shifts[m_bool_range - 1];
```
Now that I'm looking at this again, I realized I was just implementing a more complicated version of count_leading_zeroes, so I just changed it to use that instead.

Zaggy1024 commented on Userland/Libraries/LibVideo/DecoderError.h:
```diff
@@ -0,0 +1,106 @@
+/*
```
Yeah, I figure it's better to separate those for a bit more of a clear changelog, but that's ultimately up to you reviewers. :)

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/Decoder.cpp:
```diff
@@ -152,24 +215,1115 @@ u8 Decoder::adapt_prob(u8 prob, u8 counts[2])
+
+    // 1. Set a variable angle2 equal to angle & 127.
+    angle &= 127;
+    VERIFY(angle > 0);
+    // 2. If angle2 is greater than or equal to 0 and less than or equal to 32, return cos64_lookup[ angle2 ].
```
I think that VERIFY may have been from when the angle value was signed, although it should have been `angle >= 0` in the first place, I believe. Been a while since I wrote this code so I forget if there are any cases where the angle isn't a fixed value.

I'll remove the VERIFY.

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/Decoder.cpp:
```diff
@@ -152,24 +215,1115 @@ u8 Decoder::adapt_prob(u8 prob, u8 counts[2])
+    if (angle <= 64)
+        return -cos64_lookup[64 - angle];
+    // 4. If angle2 is greater than 64 and less than or equal to 96, return cos64_lookup[ angle2 - 64 ] * -1.
+    // The O3 compiler trips up here and doesn't seem to know that angle must be greater than 64 here.......
```
Looks like that was the case, not sure why it caused it to think it was reading out of bounds though.

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/Parser.cpp:
```diff
@@ -1411,28 +1447,216 @@ DecoderErrorOr<i32> Parser::read_coef(Token token)
+
+            for (auto ref_list = 0u; ref_list < 2; ref_list++) {
+                if (m_ref_frames[candidate_index][ref_list] == reference_frame) {
+                    // get_sub_block_mv()
```
This is just to mark where the function defined by spec begins here. Since it's only called in one place and it was pointless to create a whole function call for it, I just wrote it in-line here.

I'll update the comment to be clearer.

Zaggy1024 commented:
Sorry for the force push spam, but I've had a look over all the commits to check for any weird rebase shenanigans and caught a few other minor mistakes.

I've addressed all the comments so far (I believe), so it should be ready for more!

Zaggy1024 commented:
Not sure why GitHub thinks that there's an unresolved review comment, maybe because there was one comment I couldn't mark as resolved because it said it was outdated immediately. I had resolved that issue by cleaning up the removal of the verify_cast function from Utilities.h, so everything should be good to go if there's nothing else to address.

ADKaster commented:
There's actually 3 unresolved comments, one from davidot, and two from FireFox317. It says "1 change requested" because I haven't gone to the commits or Files changed tabs and left a review that says "approve" or "comment". I wouldn't worry about that one too much.

Your branch is actually 938 commits behind master :sweat_smile: . Generally the advice is to rebase your branch against upstream/master (SerenityOS/master). 

If you rebase on SerenityOS/master, and apply this diff, then your Test will run in Lagom:

```diff
diff --git a/Meta/Lagom/CMakeLists.txt b/Meta/Lagom/CMakeLists.txt
index a8d53a9a24..b7325c8fec 100644
--- a/Meta/Lagom/CMakeLists.txt
+++ b/Meta/Lagom/CMakeLists.txt
@@ -541,6 +541,13 @@ if (BUILD_LAGOM)
     )
     target_compile_definitions(LibUnicode PRIVATE ENABLE_UNICODE_DATA=$<BOOL:${ENABLE_UNICODE_DATABASE_DOWNLOAD}>)
 
+    file(GLOB LIBVIDEO_SOURCES CONFIGURE_DEPENDS "../../Userland/Libraries/LibVideo/*.cpp")
+    file(GLOB LIBVIDEO_SUBDIR_SOURCES CONFIGURE_DEPENDS "../../Userland/Libraries/LibVideo/*/*.cpp")
+    lagom_lib(Video video
+        SOURCES ${LIBVIDEO_SOURCES} ${LIBVIDEO_SUBDIR_SOURCES}
+        LIBS LibAudio LibIPC LibGfx
+    )
+
     # WASM
     file(GLOB LIBWASM_SOURCES CONFIGURE_DEPENDS "../../Userland/Libraries/LibWasm/*/*.cpp")
     lagom_lib(Wasm wasm
@@ -790,6 +797,12 @@ if (BUILD_LAGOM)
             lagom_test(${source} LIBS LibUnicode)
         endforeach()
 
+        file(GLOB LIBVIDEO_TEST_SOURCES CONFIGURE_DEPENDS "../../Tests/LibVideo/*.cpp")
+        foreach(source ${LIBVIDEO_TEST_SOURCES})
+            lagom_test(${source} LIBS LibVideo
+                WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/../../Tests/LibVideo)
+        endforeach()
+
         # JavaScriptTestRunner + LibTest tests
         # test-js
         add_executable(test-js_lagom
```

I confirmed that it runs with ASAN and UBSAN enabled, with the same environment variables as the Azure Lagom runners (Meta/Azure/Lagom.yml), so no hidden surprises there.

Zaggy1024 commented:
I've rebased onto master and added that change. For some reason, I'm still getting this error when trying to run the Lagom tests:

https://gist.github.com/Zaggy1024/a4e85fb83d5a848e51b0dcc0bbbca74f

I'm looking for help on the Discord server, but hopefully now that the test is added to Lagom it can be confirmed to work on CI?

ADKaster reviewed (approved):
Awesome! Thanks for sticking out through 110 review comments. This refactor of LibVideo to use more current error handling styles and addition of video playback is really cool to see. Let's land this thing and iterate on it in tree.

FireFox317 reviewed (commented):
I had one more pending comment, which I didn't send yet. If you are doing a follow-up PR, might be a good idea to fix the uninitialized member.

Zaggy1024 commented on Userland/Libraries/LibVideo/VP9/BitStream.h:
```diff
@@ -42,11 +44,13 @@ class BitStream {
-    i8 m_current_bit_position { 0 };
-    u64 m_bytes_read { 0 };
+    u8 const* m_end_ptr { nullptr };
+    u64 m_reservoir;
```
Ah, good point, I'll throw up a PR for that quick.

Zaggy1024 replied:
Here it is: https://github.com/SerenityOS/serenity/pull/15528
